### DISCUSSION TITLE
Variable Keys in Struct Literals

### DISCUSSION THREAD

**Discussion by Alucard17** [google/cel-go#868] [Q&A]
I have encountered unexpected behavior with optional fields in struct literals while using CEL .
The issue revolves around how keys are interpreted in struct literals, specifically with respect to optional fields.

```go 
env, err := cel.NewEnv(
cel.Variable("name", cel.StringType),
cel.Variable("group", cel.StringType),
)
ast, issues := env.Compile(`{name: 1}`)
if issues != nil && issues.Err() != nil {
log.Fatalf("type-check error: %s", issues.Err())
}
prg, err := env.Program(ast)
if err != nil {
log.Fatalf("program construction error: %s", err)
}
out, _, err := prg.Eval(map[string]interface{}{
"name": "/groups/acme.co/documents/secret-stuff",
"group": "acme.co"})
fmt.Println(out)
```
I get following output
```
{/groups/acme.co/documents/secret-stuff: 1}
```
1. `{name: 1}`: In this case, the key name is treated as an identifier, leading to its resolution from the environment variable. The resulting output is `{/groups/acme.co/documents/secret-stuff: 1}`.
2. `{[name]: 1}`: returns `{[/groups/acme.co/documents/secret-stuff]: 1}`.
3. `{["name"]: 1}`: In this example, ["name"] is explicitly quoted, indicating a constant string key. This results in `{[name]: 1}`, it surrounds the key with square brackets.

The behavior raises concerns about ambiguity when wanting to use a variable as a key. While CEL treats unquoted keys as identifiers and quoted keys as constant strings, it might be beneficial to have clearer guidelines or options for handling such scenarios.

**Comment by TristonianJones** [comment]
@Alucard17, just to be clear, `name` is not an optional field in this example. Rather, this is a variable key. An optional would look like `{?name: optional.of(1)}`.

Regarding your concern, for the variable `name`, only the types `string`, `int`, `uint`, and `bool` should be supported. Only the first example `{name: 1}` should be semantically valid. The other scenarios are bugs which we need to address both in the type-checker and the runtime. For reference, this is what the spec says: https://github.com/google/cel-spec/blob/master/doc/langdef.md#values

In terms of the printed output, the format call relies on the `String()` method which is only intended for debugging purposes.
